c-api/dotnet/src/Picky/Pem.cs

```diff
@@ -0,0 +1,170 @@
+                    throw new PemException(Error.Last());
+                }
+
+                this.label = label;
+                this.data = data;
+            }
+        }
+    }
+
+    public Pem(String pem_repr)
+    {
+        byte[] input = Utils.StringToUtf8WithNulTerminator(pem_repr);
+
+        unsafe
+        {
+            // Pin the buffer to a fixed location in memory so that garbage collector
+            // don't move it while it's being used inside the native call.
+            fixed (byte* inputPtr = input)
+            {
+                this.inner = Native.Raw.pem_parse((sbyte*)inputPtr, input.Length);
+
+                if (this.inner == null)
+                {
+                    throw new PemException(Error.Last());
+                }
+            }
+        }
+    }
+
+    public String ToRepr()
+    {
+        unsafe
+        {
+            int len = Native.Raw.pem_compute_repr_length(this.inner);
+
+            if (len == -1)
+            {
+                throw new PemException(Error.Last());
+            }
+
+            byte[] buf = new byte[len];
+
+            fixed (byte* bufPtr = buf)
+            {
+                if (Native.Raw.pem_to_repr(this.inner, (sbyte*)bufPtr, buf.Length) != len)
+                {
+                    throw new PemException(Error.Last());
+                }
+            }
+
+            return Utils.Utf8WithNulTerminatorToString(buf);
+        }
+    }
+
+    public String GetLabel()
+    {
+        if (this.label != null)
+        {
+            return this.label;
+        }
+
+        unsafe
+        {
+            int len = Native.Raw.pem_label_length(this.inner);
+
+            if (len == -1)
+            {
+                throw new PemException(Error.Last());
+            }
+
+            byte[] buf = new byte[len];
+
+            fixed (byte* bufPtr = buf)
+            {
+                if (Native.Raw.pem_label(this.inner, (sbyte*)bufPtr, buf.Length) != len)
+                {
+                    throw new PemException(Error.Last());
+                }
+            }
+
+            this.label = Utils.Utf8WithNulTerminatorToString(buf);
+
+            return this.label;
+        }
+    }
+
+    public byte[] GetData()
+    {
+        if (this.data != null)
+        {
+            return this.data;
+        }
+
+        unsafe
+        {
+            int len = Native.Raw.pem_data_length(this.inner);
+
+            if (len == -1)
+            {
+                throw new PemException(Error.Last());
+            }
+
+            byte[] buf = new byte[len];
+
+            fixed (byte* bufPtr = buf)
+            {
+                if (Native.Raw.pem_data(this.inner, bufPtr, buf.Length) != len)
+                {
+                    throw new PemException(Error.Last());
+                }
+            }
+
+            this.data = buf;
+
+            return buf;
+        }
+    }
+
+    ~Pem()
+    {
+        unsafe
+        {
+            Native.Raw.pem_drop(this.inner);
```

Comment:
I don't know the semantics behind the native object; do you want this to be deterministic? Your native object won't be dropped until the GC runs. If you want it to be deterministic, you can implement `IDisposable` - the end-user can then use `try`/`finally` or `using` or just call `Dispose` to get rid of the native reference *immediately*.

Reply:
It's just a `picky::pem::Pem` on the heap, as long as it gets dropped at some point it's fine. I figured it was more ergonomic from a C# user point of view to just let the GC do the job here, but what do you think?

Reply:
@CBenoit Yes it's totally fine as long as it doesn't need to be deterministic. I don't know the internals of `Pem` so I wasn't sure!